File: packages/amplify_storage_s3/android/src/main/kotlin/com/amazonaws/amplify/amplify_storage_s3/AmplifyStorageOperations.kt
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_storage_s3

import androidx.annotation.NonNull
import android.os.Handler
import android.os.Looper
import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.ExceptionUtil
import com.amplifyframework.core.Amplify
import io.flutter.plugin.common.MethodChannel
import java.lang.Exception
import java.text.SimpleDateFormat
import com.amazonaws.amplify.amplify_storage_s3.types.*
import com.amplifyframework.storage.StorageException
import com.amplifyframework.storage.result.*

class AmplifyStorageOperations {

    companion object StorageOperations {

        private val LOG = Amplify.Logging.forNamespace("amplify:flutter:storage_s3")

        fun uploadFile(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
            try {
                var responseSent = false
                FlutterUploadFileRequest.validate(request)
                val req = FlutterUploadFileRequest(request)
                Amplify.Storage.uploadFile(
                        req.key,
                        req.file,
                        req.options,
                        { result ->
                            prepareUploadFileResponse(flutterResult, result)
                        },
                        { error ->
                            if (!responseSent) {
                                responseSent = true
                                prepareError(flutterResult, error)
                            } else {
                                LOG.error("StorageException", error)
                            }
                        })
            } catch(e: Exception) {
                prepareError(flutterResult, e)
            }
        }

        fun getUrl(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
            try {
                FlutterGetUrlRequest.validate(request)
                val req = FlutterGetUrlRequest(request)
                Amplify.Storage.getUrl(req.key,
                        req.options,
                        { result ->
                            prepareGetUrlResponse(flutterResult, result)
                        },
                        { error ->
                            prepareError(flutterResult, error)
                        }
                )
            } catch(e: Exception) {
                prepareError(flutterResult, e)
            }
        }

        fun remove(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
            try {
                FlutterRemoveRequest.validate(request)
                val req = FlutterRemoveRequest(request)
                Amplify.Storage.remove(req.key,
                        req.options,
                        { result ->
                            prepareRemoveResponse(flutterResult, result)
                        },
                        { error ->
                            prepareError(flutterResult, error)
                        }
                )
            } catch(e: Exception) {
                prepareError(flutterResult, e)
            }
        }

        fun list(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
            try {
                FlutterListRequest.validate(request)
                val req = FlutterListRequest(request)
                Amplify.Storage.list(req.path,
                        req.options,
                        { result ->
                            prepareListResponse(flutterResult, result)
                        },
                        { error ->
                            prepareError(flutterResult, error)
                        }
                )
            } catch(e: Exception) {
                prepareError(flutterResult, e)
            }
        }

        fun downloadFile(@NonNull flutterResult: MethodChannel.Result, @NonNull request: Map<String, *>) {
            try {
                var responseSent = false
                FlutterDownloadFileRequest.validate(request)
                val req = FlutterDownloadFileRequest(request)
                Amplify.Storage.downloadFile(req.key,
                        req.file,
                        req.options,
                        { result ->
                            prepareDownloadFileResponse(flutterResult, result)
                        },
                        { error ->
                            if (!responseSent) {
                                responseSent = true
                                prepareError(flutterResult, error)
                            } else {
                                LOG.error("StorageException", error)
                            }
                        }
                )
            } catch(e: Exception) {
                prepareError(flutterResult, e)
            }
        }

        private fun prepareUploadFileResponse(@NonNull flutterResult: MethodChannel.Result, result: StorageUploadFileResult) {
            val response = HashMap<String, Any>()
            response["key"] = result.key
            Handler(Looper.getMainLooper()).post {
                flutterResult.success(response)
            }
        }

        private fun prepareGetUrlResponse(@NonNull flutterResult: MethodChannel.Result, result: StorageGetUrlResult) {
            val response = HashMap<String, Any>()
            response["url"] = result.url.toString()
            Handler(Looper.getMainLooper()).post {
                flutterResult.success(response)
            }
        }

        private fun prepareRemoveResponse(@NonNull flutterResult: MethodChannel.Result, result: StorageRemoveResult) {
            val response = HashMap<String, String>()
            response["key"] = result.key
            Handler(Looper.getMainLooper()).post {
                flutterResult.success(response)
            }
        }

        private fun prepareListResponse(@NonNull flutterResult: MethodChannel.Result, result: StorageListResult) {
            val spf = SimpleDateFormat("yyyy-MM-dd hh:mm:ss")
            val storageItemList: ArrayList<Map<String, Any>> = arrayListOf()
            for (item in result.items) {
                val storageItemMap: Map<String, Any> = mapOf(
                        "key" to item.key,
                        "eTag" to item.eTag,
                        "size" to item.size,
                        "lastModified" to spf.format(item.lastModified)
                )
                storageItemList.add(storageItemMap)
            }
            val listResultMap: Map<String, Any> = mapOf(
                    "items" to storageItemList
            )
            Handler(Looper.getMainLooper()).post {
                flutterResult.success(listResultMap)
            }
        }

        private fun prepareDownloadFileResponse(@NonNull flutterResult: MethodChannel.Result, result: StorageDownloadFileResult) {
            val response = HashMap<String, Any>()
            response["path"] = result.file.absolutePath
            Handler(Looper.getMainLooper()).post {
                flutterResult.success(response)
            }
        }

        private fun prepareError(@NonNull flutterResult: MethodChannel.Result, @NonNull e: Exception) {
            val errorCode = "StorageException"
            LOG.error(errorCode, e)
            val serializedError: Map<String, Any?> = if (e is StorageException) {
                ExceptionUtil.createSerializedError(e)
            }  else {
                ExceptionUtil.createSerializedError(
                        ExceptionMessages.missingExceptionMessage,
                        ExceptionMessages.missingRecoverySuggestion,
                        e.toString())
            }
            Handler(Looper.getMainLooper()).post {
                ExceptionUtil.postExceptionToFlutterChannel(flutterResult,
                        errorCode,
                        serializedError)
            }
        }
    }
}

File: packages/amplify_storage_s3/ios/Classes/AmplifyStorageOperations.swift
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Flutter
import UIKit
import Amplify
import AmplifyPlugins
import AWSMobileClient
import amplify_core

public class AmplifyStorageOperations {
    public static func uploadFile(flutterResult: @escaping FlutterResult, request: Dictionary<String, AnyObject>) {
        do {
            try FlutterUploadFileRequest.isValid(request: request)
            let req = FlutterUploadFileRequest(request: request)
            _ = Amplify.Storage.uploadFile(
                key: req.key,
                local: req.file,
                options: req.options,
                resultListener: { event in
                    switch event {
                    case .success(let key):
                        var result = [String: String]()
                        result["key"] = key
                        flutterResult(result)
                    case .failure(let storageError):
                        prepareError(flutterResult: flutterResult, storageError: storageError)
                    }
            })
        } catch {
            prepareGenericError(flutterResult: flutterResult, error: error)
        }
    }
    
    public static func getURL(flutterResult: @escaping FlutterResult, request: Dictionary<String, AnyObject>){
        do {
            try FlutterGetUrlRequest.isValid(request: request)
            let req = FlutterGetUrlRequest(request: request)
            _ = Amplify.Storage.getURL(key: req.key,
               options: req.options,
               resultListener: { event in
                switch event {
                case .success(let url):
                    var result = [String: String]()
                    result["url"] = url.absoluteString
                    flutterResult(result)
                case .failure(let storageError):
                    prepareError(flutterResult: flutterResult, storageError: storageError)
                }
            })
        } catch {
            prepareGenericError(flutterResult: flutterResult, error: error)
        }
    }
    
    public static func remove(flutterResult: @escaping FlutterResult, request: Dictionary<String, AnyObject>){
        do {
            try FlutterRemoveRequest.isValid(request: request)
            let req = FlutterRemoveRequest(request: request)
            _ = Amplify.Storage.remove(key: req.key,
               options: req.options,
               resultListener: { event in
                switch event {
                case .success(let key):
                    var result = [String: String]()
                    result["key"] = key
                    flutterResult(result)
                case .failure(let storageError):
                    prepareError(flutterResult: flutterResult, storageError: storageError)
                }
            })
        } catch {
            prepareGenericError(flutterResult: flutterResult, error: error)
        }
    }
    
    public static func list(flutterResult: @escaping FlutterResult, request: Dictionary<String, AnyObject>){
        do {
            try FlutterListRequest.isValid(request: request)
            let req = FlutterListRequest(request: request)
            _ = Amplify.Storage.list(options: req.options,
                 resultListener: {event in
                    switch event{
                    case .success(let result):
                        var listResultDictionary = [String:Any]();
                        var storageItemList = [[String:Any]]();
                        for item in result.items {
                            let storageItemDictionary = getStorageItemDictionary(item: item)
                            storageItemList.append(storageItemDictionary)
                        }
                        listResultDictionary["items"] = storageItemList
                        flutterResult(listResultDictionary);
                    case.failure(let storageError):
                        prepareError(flutterResult: flutterResult, storageError: storageError)
                    }
             })
        } catch {
            prepareGenericError(flutterResult: flutterResult, error: error)
        }
    }
    
    public static func downloadFile(flutterResult: @escaping FlutterResult, request: Dictionary<String, AnyObject>){
        do {
            try FlutterDownloadFileRequest.isValid(request: request)
            let req = FlutterDownloadFileRequest(request: request)
            _ = Amplify.Storage.downloadFile(key: req.key,
                 local: req.file,
                 options: req.options,
                 resultListener: { event in
                    switch event {
                    case .success:
                        var result = [String: String]()
                        // Amplify Android sends this back
                        result["path"] = req.file.absoluteURL.path
                        flutterResult(result)
                    case .failure(let storageError):
                        prepareError(flutterResult: flutterResult, storageError: storageError)
                    }
            })
        } catch {
            prepareGenericError(flutterResult: flutterResult, error: error)
        }
    }
    
    private static func getStorageItemDictionary(item: StorageListResult.Item) -> Dictionary<String ,Any>{
        let itemAsDictionary: [String : Any] = [
            "key": item.key,
            "eTag": item.eTag as Any,
            "lastModified": item.lastModified?.description as Any,
            "size": item.size as Any
        ]
        return itemAsDictionary
    }
    
    private static func prepareError(flutterResult: @escaping FlutterResult, storageError: StorageError) {
        let details: Dictionary<String, String> = createSerializedError(storageError: storageError);
        logErrorContents(error: storageError)
           ErrorUtil.postErrorToFlutterChannel(
               result: flutterResult,
               errorCode: "StorageException",
               details: details)
    }
    
    private static func prepareGenericError(flutterResult: @escaping FlutterResult, error: Error) {
        logErrorContents(error: error)
        var underlyingErrorString: String = error.localizedDescription

        if (error is AmplifyFlutterValidationException) {
                    let e = error as! AmplifyFlutterValidationException
                    underlyingErrorString = e.errorDescription
                }
                let serializedErrror = createSerializedError(
                    message: "An unexpected error has occurred",
                    recoverySuggestion: "See iOS logs for details",
                    underlyingError: underlyingErrorString)

                ErrorUtil.postErrorToFlutterChannel(
                    result: flutterResult,
                    errorCode: "StorageException",
                    details: serializedErrror)
            

    }

   static func createSerializedError(storageError: StorageError)  -> Dictionary<String, String> {
        var serializedException: Dictionary<String, String> = [:]
        serializedException["message"] = storageError.errorDescription
        serializedException["recoverySuggestion"] = storageError.recoverySuggestion
        serializedException["underlyingException"] = storageError.underlyingError.debugDescription
        return serializedException
    }

    static func createSerializedError(message: String, recoverySuggestion: String, underlyingError: String)  -> Dictionary<String, String> {
        var serializedException: Dictionary<String, String> = [:]
        serializedException["message"] = message
        serializedException["recoverySuggestion"] = recoverySuggestion
        serializedException["underlyingException"] = underlyingError
        return serializedException
    }
    
    static func logErrorContents(error: Error) {
        log.error(error: error)
    }
}

extension AmplifyStorageOperations: DefaultLogger { }
